src/lib/secrets/keypair.ml

```diff
@@ -28,33 +28,43 @@ let write_exn {Keypair.private_key; public_key} ~(privkey_path : string)
   with
   | Ok () ->
       (* The hope is that if [Secret_file.write] succeeded then this ought to
-      as well, letting [handle_open] stay inside [Secret_file]. It might not
-      if the environment changes underneath us, and we won't have nice errors
-      in that case. *)
+       as well, letting [handle_open] stay inside [Secret_file]. It might not
+       if the environment changes underneath us, and we won't have nice errors
+       in that case. *)
       let%bind pubkey_f = Writer.open_file (privkey_path ^ ".pub") in
       Writer.write_bytes pubkey_f pubkey_bytes ;
       Writer.close pubkey_f
   | Error e ->
       raise (Error.to_exn e)
 
 (** Reads a private key from [privkey_path] using [Secret_file] *)
+let read ~(privkey_path : string) ~(password : Secret_file.password) :
+    Keypair.t Deferred.Or_error.t =
+  let open Deferred.Or_error.Let_syntax in
+  let%bind pk_bytes = Secret_file.read ~path:privkey_path ~password in
+  let open Or_error.Let_syntax in
+  Deferred.return
+  @@ let%bind pk =
+       try
+         return (pk_bytes |> Bigstring.of_bytes |> Private_key.of_bigstring_exn)
+       with exn ->
+         Or_error.errorf
+           "Error parsing decrypted private key file, is your keyfile \
+            corrupt? %s"
+           (Exn.to_string exn)
+     in
+     try return (Keypair.of_private_key_exn pk)
```

Comment:
I think it makes sense to have a `Keypair.of_private_key` so that you don't have this awful `try with` statement.

Reply:
unfortunately this requires making modifications in a lot of files and digging deep through snarky source, I'll make an issue for it though